controller-server/src/main/java/com/yahoo/vespa/hosted/controller/ApplicationController.java

```diff
@@ -445,6 +445,11 @@ public LockedApplication storeWithUpdatedConfig(LockedApplication application, A
         // Validate new deployment spec thoroughly before storing it.
         controller.jobController().deploymentStatus(application.get());
 
+        // Clear notifications for instances that are no longer declared
+        for (var name : existingInstances)
+            if ( ! declaredInstances.contains(name))
+                controller.notificationsDb().removeNotifications(NotificationSource.from(application.get().id().instance(name)));
```

Comment:
We should probably also do the other cleanup, like removing DNS entries?

Reply:
Yes, I'm not sure what happens to those now and if its safe to do it here? FYI: @mpolden

Reply:
`deleteInstance(ApplicationId)` already removes DNS records.

Reply:
Yes... The problem here is that when an instance is removed from `deployment.xml`, `deleteInstance(ApplicationId)` is never called.

Reply:
😞 we should make it one shared code path.